apispec-model/src/main/scala/sttp/apispec/validation/SchemaComparator.scala

```diff
@@ -0,0 +1,534 @@
+   * @param readerNamedSchemas named schemas which may be referred to by the reader schema
+   * @return a list of incompatibilities between the schemas
+   */
+  def compare(
+    writerSchema: Schema,
+    readerSchema: Schema,
+    writerNamedSchemas: Map[String, Schema] = Map.empty,
+    readerNamedSchemas: Map[String, Schema] = Map.empty
+  ): List[SchemaCompatibilityIssue] =
+    new SchemaComparator(writerNamedSchemas, readerNamedSchemas)
+      .compare(writerSchema, readerSchema)
+}
+
+private class SchemaComparator(
+  writerNamedSchemas: Map[String, Schema],
+  readerNamedSchemas: Map[String, Schema]
+) {
+
+  import SchemaComparator._
+
+  private val cache = new mutable.HashMap[(Schema, Schema), List[SchemaCompatibilityIssue]]
+
+  // keeps schema pairs for which comparison is ongoing in order to short circuit recursive schema comparisons
+  private val inComparison = new mutable.HashSet[(Schema, Schema)]
+
+  private def compare(writerSchema: SchemaLike, readerSchema: SchemaLike): List[SchemaCompatibilityIssue] = {
+    val normalizedWriterSchema = normalize(writerSchema, writerNamedSchemas)
+    val normalizedReaderSchema = normalize(readerSchema, readerNamedSchemas)
+    val cacheKey = (normalizedWriterSchema, normalizedReaderSchema)
+    // short-circuit recursive comparison of the same pair of schemas that is already being compared
+    if (inComparison(cacheKey)) Nil
+    else cache.getOrElseUpdate(cacheKey, {
+      inComparison.add(cacheKey)
+      try compareNormalized(normalizedWriterSchema, normalizedReaderSchema)
+      finally inComparison.remove(cacheKey)
+    })
+  }
+
+  // translate AnySchema to Schema, remove annotations and resolve references
+  @tailrec private def normalize(schema: SchemaLike, named: Map[String, Schema]): Schema = schema match {
+    case AnySchema.Anything => Schema.Empty
+    case AnySchema.Nothing => Schema.Nothing
+    case s: Schema => deannotate(s) match {
+      case s@LocalReference(name) =>
+        def noSchema: Nothing =
+          throw new NoSuchElementException(s"could not resolve schema reference ${s.$ref.get}")
+
+        normalize(named.getOrElse(name, noSchema), named)
+      case s => s
+    }
+  }
+
+  private object LocalRef {
+    def unapply(ref: String): Option[String] =
+      if (ref.startsWith(RefPrefix)) Some(ref.stripPrefix(RefPrefix))
+      else None
+  }
+
+  /** Matches a schema that is a pure reference to one of the component schemas */
+  private object LocalReference {
+    def unapply(schema: Schema): Option[String] =
+      schema.$ref
+        .filter(ref => schema == Schema($ref = Some(ref)))
+        .collect {
+          case LocalRef(name) => name
+        }
+  }
+
+  // strip fields which do not affect schema comparison
+  private def deannotate(schema: Schema): Schema =
+    schema.copy(
+      $comment = None,
+      title = None,
+      description = None,
+      default = None,
```

Comment:
According to [JSON Schema spec](https://json-schema.org/draft/2020-12/json-schema-validation#name-default), the `default` keyword is an _annotation_, which is "for documentation and user interface display purposes".

Changing the default value cannot make any previously valid data invalid according to a new schema, so technically it's not an incompatibility. Whether this can be an incompatibility on a higher, semantic level is debatable. I can imagine situations where it is and where it isn't.

For example, an integer field in a request may have had a default value of 0, but then the server decided to make this field nullable, with a default value of `null`, so that it can distinguish between passing 0 explicitly and not passing anything.

Reply:
Ok, sounds reasonable, thanks for the explanation :)